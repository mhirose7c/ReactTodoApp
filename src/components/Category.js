import React from "react";
import Db from "../db";
import Task from "./Task";

const DEFAULT_PLACEHOLDER = "カテゴリーの追加";

class CategoryItems extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      rows: [],
      inputTarget: null,
      categoryName: "",
    };
    this.add = this.add.bind(this);
    this.update = this.updateCategory.bind(this);
    this.delete = this.delete.bind(this);
    this.addCategory = this.addCategory.bind(this);
    this.updateCategory = this.updateCategory.bind(this);
    this.deleteCategory = this.deleteCategory.bind(this);
    this.inputCategoryName = this.inputCategoryName.bind(this);
    this.targetChange = this.targetChange.bind(this);
    this.renderAllCategories();
  }
  async renderAllCategories() {
    const categories = await Db.getAll("category");
    this.setState({ rows: categories });
  }

  async add(category_name) {
    const newItem = Object.assign({}, { name: category_name });
    const id = await Db.add("category", newItem);
    newItem.id = id;
    this.state.rows.push(newItem);
    this.setState({ rows: this.state.rows });
    this.setState({ categoryName: "" });
  }

  async update(data) {
    await Db.update("category", data);
    this.setState({ rows: this.state.rows });
    this.setState({ inputTarget: null });
    this.setState({ categoryName: "" });
  }

  async delete(id) {
    Db.delete("category", id);
    // 付随するタスクを全て削除

    const tasks = await Db.findByIndexKey("task", "category_id", id);
    tasks.forEach((task) => {
      Db.delete("task", task.id);
    });
    this.setState({ rows: [] });
    this.renderAllCategories();
  }

  addCategory(e) {
    this.add(e.target.value);
  }

  updateCategory(e) {
    const id = e.target.dataset.categoryId;
    const name = e.target.value;
    var targetdata;
    this.state.rows.forEach((item) => {
      if (item.id == id) {
        item.name = name;
        targetdata = item;
      }
    });
    this.update(targetdata);
  }

  deleteCategory(e) {
    const id = e.target.dataset.categoryId;
    this.delete(parseInt(id));
  }

  inputCategoryName(e) {
    this.setState({ categoryName: e.target.value });
  }

  targetChange(e) {
    this.setState({ inputTarget: e.target.dataset.categoryId });
    this.setState({ categoryName: e.target.dataset.categoryName });
  }

  render() {
    let categoryItems;
    let addCategory;
    // 編集時はカテゴリの追加を非表示にする
    if (this.state.inputTarget === null) {
      addCategory = (
        <li className="category-item">
          <input
            placeholder={DEFAULT_PLACEHOLDER}
            value={this.state.categoryName}
            onChange={this.inputCategoryName}
            onBlur={this.addCategory}
          />
        </li>
      );
    }
    categoryItems = (
      <ul>
        {this.state.rows.map((category) => {
          if (this.state.inputTarget == category.id) {
            return (
              <li className="category-item">
                <input
                  data-category-id={category.id}
                  placeholder={DEFAULT_PLACEHOLDER}
                  value={this.state.categoryName}
                  onChange={this.inputCategoryName}
                  onBlur={this.updateCategory}
                />
                <Task category_id={category.id} />
              </li>
            );
          } else {
            return (
              <li className="category-item">
                <p className="category-label-wrapper">
                  <p
                    data-category-id={category.id}
                    data-category-name={category.name}
                    className="category-label"
                    onClick={this.targetChange}
                  >
                    {category.name}
                    <button
                      data-category-id={category.id}
                      className="category-config-button"
                      onClick={this.deleteCategory}
                    >
                      ×
                    </button>
                  </p>
                </p>
                <Task category_id={category.id} />
              </li>
            );
          }
        })}
        {addCategory}
      </ul>
    );
    return <div className="category-wrapper">{categoryItems}</div>;
  }
}

export default CategoryItems;
