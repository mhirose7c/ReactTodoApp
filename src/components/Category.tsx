import React, { useState, useEffect, MouseEvent } from "react";
import Db, { CategoryType, TaskType } from "../db";
import Task from "./Task";

const DEFAULT_PLACEHOLDER = "カテゴリーの追加";

export default function CategoryItems() {
  const [rows, setRows] = useState<CategoryType[]>([]);
  const [inputTarget, setInputTarget] = useState(-1);
  const [categoryName, setCategoryName] = useState("");

  useEffect(() => {
    const fetchData = async () => {
      const result = await Db.getAll("category");
      setRows(result as CategoryType[]);
    };
    fetchData();
  }, [rows]);

  function handleAddData(e: React.ChangeEvent<HTMLInputElement>) {
    const {value} = e.currentTarget
    const name: string = value;
    if (name.length == 0) {
      return;
    }
    const newItem = Object.assign({}, { name: name });
    const addData = async () => {
      const result = await Db.add("category", newItem);
      return result;
    };
    addData();
    setInputTarget(-1);
    setCategoryName("");
  }
  function parseId(target: string | undefined): number {
    return target ? parseInt(target) : -1;
  };

  function handleUpdateData(e: React.FocusEvent<HTMLInputElement>) {
    const {value, dataset} = e.currentTarget;
    const id = parseId(dataset.categoryId);
    const name = value;
    if (id == -1){
      return;
    }
    var targetdata: CategoryType;
    rows.forEach((item: CategoryType) => {
      if (item.id == id) {
        item.name = name;
        targetdata = item;
      }
    });
    const updateData = async () => {
      const result = await Db.update("category", targetdata);
      return result;
    };
    updateData();
    setInputTarget(-1);
    setCategoryName("");
  }

  function handleDeleteData(e: React.MouseEvent<HTMLElement>) {
    const id = parseId(e.currentTarget.dataset.categoryId);
    const deleteData = async () => {
      await Db.delete("category", id);
      const tasks = await Db.findByIndexKey("task", "category_id", id) as TaskType[];
      tasks.forEach((task: TaskType) => {
        Db.delete("task", task.id);
      });
    };
    deleteData();
    setInputTarget(-1);
    setCategoryName("");
  }

  function handleChangeValue(e: React.ChangeEvent<HTMLInputElement>) {
    const { value } = e.currentTarget;
    setCategoryName(value);
  }

  let categoryItems;
  let addCategory;
  // 編集時はカテゴリの追加を非表示にする
  if (inputTarget === -1) {
    addCategory = (
      <li className="category-item">
        <input
          placeholder={DEFAULT_PLACEHOLDER}
          value={categoryName}
          onChange={handleChangeValue}
          onBlur={handleAddData}
        />
      </li>
    );
  }
  categoryItems = (
    <ul>
      {rows.map((category: CategoryType) => {
        if (inputTarget == category.id) {
          return (
            <li className="category-item" key={category.id}>
              <input
                data-category-id={category.id}
                placeholder={DEFAULT_PLACEHOLDER}
                value={categoryName}
                onChange={handleChangeValue}
                onBlur={handleUpdateData}
              />
              <Task category_id={category.id} />
            </li>
          );
        } else {
          return (
            <li className="category-item" key={category.id}>
              <div className="category-label-wrapper">
                <div
                  data-category-id={category.id}
                  data-category-name={category.name}
                  className="category-label"
                  onClick={() => {
                    setInputTarget(category.id);
                    setCategoryName(category.name);
                  }}
                >
                  {category.name}
                  <button
                    data-category-id={category.id}
                    className="category-config-button"
                    onClick={handleDeleteData}
                  >
                    ×
                  </button>
                </div>
              </div>
              <Task category_id={category.id} />
            </li>
          );
        }
      })}
      {addCategory}
    </ul>
  );

  return <div className="category-wrapper">{categoryItems}</div>;
}
