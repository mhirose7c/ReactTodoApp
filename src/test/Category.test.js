import React from 'react';
import Category from '../components/Category';
import renderer from 'react-test-renderer';

jest.mock("../components//Task", () => "Task");

it('matches to the snapshot', () => {
  const tree = renderer.create(<Category />).toJSON();
  expect(tree).toMatchSnapshot()

})