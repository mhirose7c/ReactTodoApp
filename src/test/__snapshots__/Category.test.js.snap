// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches to the snapshot 1`] = `
<div
  className="category-wrapper"
>
  <ul>
    <li
      className="category-item"
    >
      <input
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="カテゴリーの追加"
        value=""
      />
    </li>
  </ul>
</div>
`;
